pkg/query/storeset.go

```diff
@@ -420,7 +448,17 @@ func (s *StoreSet) getActiveStores(ctx context.Context, stores map[string]*store
 			st, seenAlready := stores[addr]
 			if !seenAlready {
 				// New store or was unactive and was removed in the past - create new one.
-				conn, err := grpc.DialContext(ctx, addr, s.dialOpts...)
+				dialOpts := s.dialOpts
+				// Update the dialOpts if we are asked to add dnsServerName.
+				if s.dnsServerName {
```

Comment:
Hmm this seems a bit odd. The NewStoreSet func already accepts a slice of dialopts. I think we should expect that the dialopts has the correct WithTransportCredentials already set, no? ie correctly configure this option in cmd/Thanos/query rather than leaking configuration into this struct. Also, it looks to me like the server name is already set in the dialopts generated with extgrpc https://github.com/j3p0uk/thanos/blob/40526f52f54d4501737e5246c0e71e56dd7e0b2d/pkg/extgrpc/client.go#L56

Reply:
Also, taking out this new dial opt and server name bool from this struct would eliminate the need for extending the StoreSpec interface and adding the extra field to the grpcStoreSpec struct

Reply:
The bug, as it stands, is that the arguments provided to the query at start time do not allow for individual server names per configured store.  The store set gets populated with the resolved addresses of the stores in the store spec and not the original server name, as far as I could tell.  Given this, passing through the flag to generate the TLS portion of the dialOpts on demand at connection time by retrieving the server name from the dns provider.  I passed minimal data through to accomplish that.

Reply:
Indeed this is a bit weird because with this we not only have two places where these options are set but also this newly introduced option overrides the other one. It should probably be refactored so that `query.NewGRPCStoreSpec*` would hook into some function which would give it a list of opts to use. We could discuss it on #thanos-dev.
Also, I wonder if there any downsides to enabling this kind of feature by default for everyone, and then only allow overriding the server name via the option that exists already? That would probably simplify things too.

Reply:
Sure, happy to discuss further.

My thoughts are that this looks odd because a StoreSet is a set of Stores, and the code assumes that each of the stores shares a particular set of characteristics, which may not be true (and is the basis for this bug fix).  The StoreSet currently holds attributes that can differ on a per-store basis.  The ServerName is one such attribute.

Reply:
Yes, but I think we need to refactor this because the current approach causes the StoreSet to have duplicate `WithTransportCredentials` dialopts, one of which overrides the other. This makes it hard to reason about the code IMO

Reply:
Chat in slack...

Sounds like implementation of this would be something like removing dialOpts from the StoreSet, and adding it to the StoreSpec.
We should be able to generate the dialOpts at creation of the StoreSpec array in the StoreSet, so would only be once at startup.
Need to decide on what flags should exist, and how would they interact and how to handle address-specified stores and DNS SRV records.

For the SRV case, I think the correct behaviour would be to enable lookup of the A record resolved from the SRV lookup.  This data is currently lost during resolution, as only the IPs are returned to the provider from resolver, and stored under the key of the original store spec.
To improve the DNS Provider lookup of address to server name, we could pass back the data to stash in the provider as a reverse lookup cache.  This would allow SRV to do addr -> A lookup, and others to do addr -> host lookup, simplifying the Provider.ServerName() function.

I can drop the “--grpc-client-dns-server-name” flag and then use “dns” in the “--grpc-client-server-name” to trigger addition of server name from DNS in the dialOpts, as “dns” is likely an illegal server name anyway.

Reply:
@bwplotka @squat ping ^